docs/adr/0007-use-structured-proxy-types.md

```diff
@@ -0,0 +1,28 @@
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+
+- [7. Use structured proxy types](#7-use-structured-proxy-types)
+  - [Context](#context)
+  - [Decision](#decision)
+  - [Consequences](#consequences)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+# 7. Use structured proxy types
+
+- Status: proposed
+- Date: 2023-08-09
+- Authors: @Danil-Grigorev
+- Deciders: @richardcase @alexander-demicev @salasberryfin @Danil-Grigorev @mjura
+
+## Context
+
+As described in `Rancher` integration strategy [ADR](./0005-rancher-integration-strategy.md#context) we'd like to operate on several `CRs` provided by the `Rancher`. Although the use of unstructured objects may be avoided by specifying a subset definition of the `CRs` we are using, without a need to import `Rancher` API packages directly.
+
+## Decision
+
+The `rancher-turtles` operator will be using `kubebuilder` [annotations](https://kubebuilder.io/reference/markers) on specified for the `Rancher` proxy types located under the `./internal/rancher` directory, to generate deep copy definitions and therefore allow specified resources to match the `Object` [interface](https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/client/object.go#L45) provided by the controller-runtime. This will allow us to operate independently from `Rancher` specified types, without sacrificing functionality provided by the `Object` interface, including but not limited to a variety of helper methods. This also allows us to avoid unnecessary overhead in converting our objects to unstructured, inconsistent behaviors happening around setting API types on these objects and allows us to use the `envtest` and avoid using the `fakeClient` while testing (see the [issue](https://github.com/kubernetes-sigs/controller-runtime/issues/2308)) and use regular [setup procedures](https://kubebuilder.io/reference/envtest.html?highlight=envtest#writing-tests) for envtest around type registration. This will also allow us to simplify the process of writing e2e [tests](../../test/e2e/README.md#testing) and would allow us to know in advance if the specification of proxy type diverges from `Rancher` spec, using the latest `Rancher` version in e2e.
```

Comment:
Can we make a list of "allows us features" with bullet points to avoid repetition?